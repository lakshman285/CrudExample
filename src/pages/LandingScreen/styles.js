import {StyleSheet} from 'react-native';
import {colors} from '../../typography/theme';
import {fonts} from '../../typography/fonts';

const Styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.white,
  },
  contentContainer: {flexGrow: 1},
  containerView: {
    display: 'flex',
    flexDirection: 'column',
    flex: 1,
    justifyContent: 'space-between',
  },
  topContainerView: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'flex-start',
  },
  logoContainer: {
    paddingTop: 75,
  },
  logoImageContainer: {
    justifyContent: 'center',
    alignItems: 'center',
    padding: 30,
  },
  logo: {
    width: 90,
    height: 90,
  },
  signInContainer: {
    padding: 0,
    alignItems: 'center',
  },
  signInText: {
    textAlign: 'center',
    fontSize: fonts.fontSize.lg,
    fontWeight: 'bold',
    color: colors.black,
    width: 250,
  },
  signInDescriptionText: {
    textAlign: 'center',
    fontSize: fonts.fontSize.md,
    color: colors.black,
    marginTop: 10,
    width: 250,
  },
  bottomContainerView: {
    padding: 20,
  },
  appVersion: {
    textAlign: 'center',
    fontSize: fonts.fontSize.sm,
    fontWeight: 'bold',
    textTransform: 'uppercase',
  },
});

export default Styles;
