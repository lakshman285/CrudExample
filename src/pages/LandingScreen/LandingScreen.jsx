import {
  View,
  Text,
  ScrollView,
  Image,
  SafeAreaView,
  Platform,
  Alert,
  AppState,
} from 'react-native';
import React from 'react';
import styles from './styles';
import {name as app_name, version as app_version} from '../../../package.json';
import {useEffect, useState} from 'react';
import {
  check,
  request,
  openSettings,
  PERMISSIONS,
  RESULTS,
} from 'react-native-permissions';
import Geolocation from '@react-native-community/geolocation';
import {useNavigation} from '@react-navigation/native';
import {isAuthenticated} from '@okta/okta-react-native';

const LandingScreen = () => {
  const navigation = useNavigation();
  const [appState, setAppState] = useState(AppState.currentState);
  const [hasReturnedFromBackground, setHasReturnedFromBackground] =
    useState(false);
  const [hasClickedOnSettings, setHasClickedOnSettings] = useState(false);

  useEffect(() => {
    checkUserLocationAccess();
  }, []);

  useEffect(() => {
    //Event Listener for changes in app state
    const handleAppStateChange = nextAppState => {
      if (
        appState === 'background' &&
        nextAppState === 'active' &&
        hasClickedOnSettings
      ) {
        //The app has returned from the background
        setHasReturnedFromBackground(true);
      }
      setAppState(nextAppState);
    };

    //subscribe to app state changes
    const subscription = AppState.addEventListener(
      'change',
      handleAppStateChange,
    );

    //cleanup when component unmounts
    return () => {
      subscription.remove();
    };
  }, [appState]);

  useEffect(() => {
    if (hasReturnedFromBackground) {
      //Do something when the app returns from the background
      checkUserLocationAccess();
      setHasReturnedFromBackground(false); //reset the flag
      setHasClickedOnSettings(false);
    }
  }, [hasReturnedFromBackground]);

  /**
   * This function is used to check whether user enabled location permission or not
   */
  const checkUserLocationAccess = () => {
    //for android
    if (Platform.OS === 'android') {
      check(PERMISSIONS.ANDROID.ACCESS_FINE_LOCATION).then(result => {
        if (result == RESULTS.GRANTED) {
          getUserLocation();
        } else {
          requestLocationPermission();
        }
      });
    } else {
      //for ios
      check(PERMISSIONS.IOS.LOCATION_WHEN_IN_USE).then(result => {
        if (result == RESULTS.GRANTED) {
          getUserLocation();
        } else {
          requestLocationPermission();
        }
      });
    }
  };

  /**
   * This function is used to display dialog when user denied location permission
   * @returns location permission denied dialog
   */
  const displayPermissionDeniedDialog = () => {
    return Alert.alert(
      'Require Location Permission',
      'App need location access to prompt you to check-in and out from app',
      [
        {
          text: 'OK',
          onPress: () => {
            checkUserLocationAccess();
          },
        },
      ],
      {cancelable: false},
    );
  };

  /**
   * This function is used to display dialog when user permanently denied/blocked location permission
   * @returns location permission permanently denied/blocked dialog
   */
  const displayOpenSettingsDialog = () => {
    return Alert.alert(
      'Require Location Permission',
      'App need location access to prompt you to check-in and out from app',
      [
        {
          text: 'Open Settings',
          onPress: () => {
            openSettings();
            setHasClickedOnSettings(true);
          },
        },
      ],
      {cancelable: false},
    );
  };

  /**
   * This function is used to report location permission
   */
  const requestLocationPermission = () => {
    //for android
    if (Platform.OS === 'android') {
      request(PERMISSIONS.ANDROID.ACCESS_FINE_LOCATION).then(result => {
        if (result == RESULTS.GRANTED) {
          getUserLocation();
        } else if (result == RESULTS.DENIED) {
          displayPermissionDeniedDialog();
        } else if (result == RESULTS.BLOCKED) {
          displayOpenSettingsDialog();
        }
      });
    } else {
      //for ios
      request(PERMISSIONS.IOS.LOCATION_WHEN_IN_USE).then(result => {
        if (result == RESULTS.GRANTED) {
          getUserLocation();
        } else if (result == RESULTS.DENIED) {
          displayPermissionDeniedDialog();
        } else if (result == RESULTS.BLOCKED) {
          displayOpenSettingsDialog();
        }
      });
    }
  };

  /**
   * This method is used to get user location
   */
  const getUserLocation = () => {
    Geolocation.getCurrentPosition(
      //will give you the current location
      position => {
        //getting the longitude from the location json
        const currentLongitude = JSON.stringify(position.coords.longitude);

        //getting the latitude from the location json
        const currentLatitude = JSON.stringify(position.coords.latitude);

        getPostalAddress();
      },
      error => {
        console.log(
          'Get One Time Location permission denied',
          JSON.stringify(error),
        );
      },
      {
        enableHighAccuracy: false,
        timeout: 300000,
        maximumAge: 1000,
      },
    );
  };

  const getPostalAddress = async () => {
    console.log('Getting Postal Address');
    checkAuthentication();
  };

  /**
   * This Method is used to check whether user authenticated or not
   * @todo navigate to actual screen when user logged in
   */
  const checkAuthentication = async () => {
    const result = await isAuthenticated();
    if (result.authenticated) {
      navigation.replace('signInLoad');
    } else {
      navigation.replace('Home');
    }
  };

  return (
    <SafeAreaView style={styles.container}>
      <ScrollView contentContainerStyle={styles.contentContainer}>
        <View style={styles.containerView}>
          <View style={styles.topContainerView}>
            <View style={styles.logoContainer}>
              <View style={styles.logoImageContainer}>
                <Image style={styles.logo} source={require('')} />
              </View>
            </View>
            <View style={styles.signInContainer}>
              <Text style={styles.signInText}>Sign In</Text>
              <Text style={styles.signInDescriptionText}>
                Crud Example Needs Sign In
              </Text>
            </View>
          </View>
          <View style={styles.bottomContainerView}>
            <Text style={styles.appVersion}>
              {app_name} {app_version}
            </Text>
          </View>
        </View>
      </ScrollView>
    </SafeAreaView>
  );
};

export default LandingScreen;
