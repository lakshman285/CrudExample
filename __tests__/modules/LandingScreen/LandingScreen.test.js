import React from 'react';
import renderer from 'react-test-renderer';
import LandingScreen from '../../../src/pages/LandingScreen/LandingScreen';
import {render, waitFor} from '@testing-library/react-native';
import {NavigationContainer} from '@react-navigation/native';
import {Platform} from 'react-native';
import {request, check} from 'react-native-permissions';

jest.mock('@okta/okta-react-native', () => ({
  isAuthenticated: jest.fn(() => Promise.resolve({authenticated: true})),
}));

jest.mock('react-native-safe-area-context', () => {
  const inset = {top: 0, right: 0, bottom: 0, left: 0};
  return {
    SafeAreaProvider: jest.fn().mockImplementation(({children}) => children),
    SafeAreaConsumer: jest
      .fn()
      .mockImplementation(({children}) => children(inset)),
    useSafeAreaInsets: jest.fn().mockImplementation(() => inset),
  };
});

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: jest.fn(),
      setOptions: jest.fn(),
      getState: jest.fn(),
      goBack: jest.fn(),
      replace: jest.fn(),
    }),
  };
});

describe('Landing Screen Page renders correctly', () => {
  it('Renders the Landing Screen Correctly', () => {
    const tree = renderer
      .create(
        <NavigationContainer>
          <LandingScreen />
        </NavigationContainer>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  const navigation = {navigate: () => jest.fn()};

  const test = (
    <NavigationContainer>
      <LandingScreen navigation={navigation} />
    </NavigationContainer>
  );

  it('should render Landing Screen Page for Android', async () => {
    Platform.OS = 'android';
    const wrapper = render(test);
    await waitFor(() => {
      expect(wrapper).toBeTruthy();
    });
  });

  it('should render Landing Screen Page for IOS', async () => {
    Platform.OS = 'ios';
    const wrapper = render(test);
    await waitFor(() => {
      expect(wrapper).toBeTruthy();
    });
  });

  it('should render Landing Screen Page for State Change', async () => {
    const hasReturnedFromBackground = true;
    const setHasReturnedFromBackground = jest.fn();
    jest
      .spyOn(React, 'useState')
      .mockImplementation(() => [
        hasReturnedFromBackground,
        setHasReturnedFromBackground,
      ]);
    const wrapper = render(test);
    await waitFor(() => {
      expect(wrapper).toBeTruthy();
    });
  });

  it('should render check android landing page denied permission', async () => {
    Platform.OS = 'android';
    check.mockImplementation(() => new Promise(resolve => resolve('denied')));
    const wrapper = render(test);
    await waitFor(() => {
      expect(wrapper).toBeTruthy();
    });
  });

  it('should render request android landing page denied permission', async () => {
    Platform.OS = 'android';
    check.mockImplementation(() => new Promise(resolve => resolve('denied')));
    const wrapper = render(test);
    await waitFor(() => {
      expect(wrapper).toBeTruthy();
    });
  });

  it('should render check android landing page blocked permission', async () => {
    Platform.OS = 'android';
    check.mockImplementation(() => new Promise(resolve => resolve('blocked')));
    const wrapper = render(test);
    await waitFor(() => {
      expect(wrapper).toBeTruthy();
    });
  });

  it('should render request android landing page blocked permission', async () => {
    Platform.OS = 'android';
    check.mockImplementation(() => new Promise(resolve => resolve('blocked')));
    const wrapper = render(test);
    await waitFor(() => {
      expect(wrapper).toBeTruthy();
    });
  });

  it('should render check android landing page granted permission', async () => {
    Platform.OS = 'android';
    check.mockImplementation(() => new Promise(resolve => resolve('granted')));
    const wrapper = render(test);
    await waitFor(() => {
      expect(wrapper).toBeTruthy();
    });
  });

  it('should render request android landing page granted permission', async () => {
    Platform.OS = 'android';
    check.mockImplementation(() => new Promise(resolve => resolve('granted')));
    const wrapper = render(test);
    await waitFor(() => {
      expect(wrapper).toBeTruthy();
    });
  });

  it('should render check ios landing page denied permission', async () => {
    Platform.OS = 'ios';
    check.mockImplementation(() => new Promise(resolve => resolve('denied')));
    const wrapper = render(test);
    await waitFor(() => {
      expect(wrapper).toBeTruthy();
    });
  });

  it('should render request ios landing page denied permission', async () => {
    Platform.OS = 'ios';
    check.mockImplementation(() => new Promise(resolve => resolve('denied')));
    const wrapper = render(test);
    await waitFor(() => {
      expect(wrapper).toBeTruthy();
    });
  });

  it('should render check ios landing page blocked permission', async () => {
    Platform.OS = 'ios';
    check.mockImplementation(() => new Promise(resolve => resolve('blocked')));
    const wrapper = render(test);
    await waitFor(() => {
      expect(wrapper).toBeTruthy();
    });
  });

  it('should render request ios landing page blocked permission', async () => {
    Platform.OS = 'ios';
    check.mockImplementation(() => new Promise(resolve => resolve('blocked')));
    const wrapper = render(test);
    await waitFor(() => {
      expect(wrapper).toBeTruthy();
    });
  });

  it('should render check ios landing page granted permission', async () => {
    Platform.OS = 'ios';
    check.mockImplementation(() => new Promise(resolve => resolve('granted')));
    const wrapper = render(test);
    await waitFor(() => {
      expect(wrapper).toBeTruthy();
    });
  });

  it('should render request ios landing page granted permission', async () => {
    Platform.OS = 'ios';
    check.mockImplementation(() => new Promise(resolve => resolve('granted')));
    const wrapper = render(test);
    await waitFor(() => {
      expect(wrapper).toBeTruthy();
    });
  });   
});
