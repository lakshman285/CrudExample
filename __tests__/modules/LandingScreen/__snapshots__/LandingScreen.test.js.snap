// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing Screen Page renders correctly Renders the Landing Screen Correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "flexGrow": 1,
      }
    }
  >
    <View>
      <View
        style={
          {
            "display": "flex",
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "space-between",
          }
        }
      >
        <View
          style={
            {
              "display": "flex",
              "flexDirection": "column",
              "justifyContent": "flex-start",
            }
          }
        >
          <View
            style={
              {
                "paddingTop": 75,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "padding": 30,
                }
              }
            >
              <Image
                source={
                  {
                    "default": [Function],
                  }
                }
                style={
                  {
                    "height": 90,
                    "width": 90,
                  }
                }
              />
            </View>
          </View>
          <View
            style={
              {
                "alignItems": "center",
                "padding": 0,
              }
            }
          >
            <Text
              style={
                {
                  "color": "black",
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "textAlign": "center",
                  "width": 250,
                }
              }
            >
              Sign In
            </Text>
            <Text
              style={
                {
                  "color": "black",
                  "fontSize": 14,
                  "marginTop": 10,
                  "textAlign": "center",
                  "width": 250,
                }
              }
            >
              Crud Example Needs Sign In
            </Text>
          </View>
        </View>
        <View
          style={
            {
              "padding": 20,
            }
          }
        >
          <Text
            style={
              {
                "fontSize": 12,
                "fontWeight": "bold",
                "textAlign": "center",
                "textTransform": "uppercase",
              }
            }
          >
            CrudExample
             
            0.0.1
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
